src/frame.rs

```diff
@@ -0,0 +1,163 @@
+use crate::err::{ConstructionError, CanError, CanErrorDecodingFailure};
+use embedded_hal::can::{Frame, Id, StandardId, ExtendedId};
+use crate::constants::*;
+use crate::util::hal_id_to_raw;
+
+use std::fmt;
+
+use itertools::Itertools;
+
+
+/// CanFrame
+///
+/// Uses the same memory layout as the underlying kernel struct for performance
+/// reasons.
+#[derive(Default, Debug, Copy, Clone)]
+#[repr(C)]
+pub struct CanFrame {
+    /// 32 bit CAN_ID + EFF/RTR/ERR flags
+    _id: u32,
+
+    /// data length. Bytes beyond are not valid
+    _data_len: u8,
+
+    /// padding
+    _pad: u8,
+
+    /// reserved
+    _res0: u8,
+
+    /// reserved
+    _res1: u8,
+
+    /// buffer for data
+    _data: [u8; 8],
+}
+
+impl CanFrame {
+    pub fn init(id: u32, data: &[u8], rtr: bool, err: bool) -> Result<CanFrame, ConstructionError> {
+        let mut _id = id;
+
+        if data.len() > 8 {
+            return Err(ConstructionError::TooMuchData);
+        }
+
+        if id > EFF_MASK {
+            return Err(ConstructionError::IDTooLarge);
+        }
+
+        // set EFF_FLAG on large message
+        if id > SFF_MASK {
+            _id |= EFF_FLAG;
+        }
+
+
+        if rtr {
+            _id |= RTR_FLAG;
+        }
+
+        if err {
+            _id |= ERR_FLAG;
+        }
+
+        let mut full_data = [0; 8];
+
+        // not cool =/
+        for (n, c) in data.iter().enumerate() {
+            full_data[n] = *c;
+        }
+
+        Ok(CanFrame {
+               _id,
+               _data_len: data.len() as u8,
+               _pad: 0,
+               _res0: 0,
+               _res1: 0,
+               _data: full_data,
+           })
+    }
+
+    /// Return the error message
+    pub fn err(&self) -> u32 {
+        self._id & ERR_MASK
+    }
+
+    /// Check if frame is an error message
+    pub fn is_error(&self) -> bool {
+        self._id & ERR_FLAG != 0
+    }
+
+    // Read error from message and transform it into a `CanError`.
+    //
+    // SocketCAN errors are indicated using the error bit and coded inside
+    // id and data payload. Call `error()` converts these into usable
+    // `CanError` instances.
+    //
+    // If the frame is malformed, this may fail with a
+    // `CanErrorDecodingFailure`.
+    #[inline]
+    pub fn error(&self) -> Result<CanError, CanErrorDecodingFailure> {
+        CanError::from_frame(self)
+    }
+
+}
+
+impl Frame for CanFrame {
+    /// Create a new frame
+    fn new(id: impl Into<Id>, data: &[u8]) -> Option<Self> {
+        let raw_id = hal_id_to_raw(id.into());
```

Comment:
You don't appear to be properly handling when someone passes in an `ExtendedId` with a value <= SFF_MASK.

That was a known bug with the previous implementation. See #15.

We should be able to pass a unit test like:
```
let frame = CanFrame::new(ExtendedId::new(0x50), &[]).unwrap();
assert!(frame.is_extended());
```